src/mem/x86_64.rs

```diff
@@ -86,20 +114,37 @@ pub unsafe fn set_bytes(dest: *mut u8, c: u8, count: usize) {
 
 #[inline(always)]
 #[cfg(not(target_feature = "ermsb"))]
-pub unsafe fn set_bytes(dest: *mut u8, c: u8, count: usize) {
-    let qword_count = count >> 3;
-    let byte_count = count & 0b111;
-    // FIXME: Use the Intel syntax once we drop LLVM 9 support on rust-lang/rust.
-    core::arch::asm!(
-        "repe stosq %rax, (%rdi)",
-        "mov {byte_count:e}, %ecx",
-        "repe stosb %al, (%rdi)",
-        byte_count = in(reg) byte_count,
+pub unsafe fn set_bytes(mut dest: *mut u8, c: u8, count: usize) {
+    let c = c as u64 * 0x0101_0101_0101_0101;
+    let (pre_byte_count, qword_count, byte_count) = rep_param(dest, count);
+    // Separating the blocks gives the compiler more freedom to reorder instructions.
+    // It also allows us to trivially skip the rep stosb, which is faster when memcpying
+    // aligned data.
+    if pre_byte_count > 0 {
```

Comment:
I would expect a `rep stosb` with an `ecx` value of 0 to act like a no-op anyways. Is there a perf benefit to keeping the `if` here?

Reply:
There is a measurable benefit for `memcpy_rust_4096` on my machine at least:

with branch:
```
test memcpy_rust_1048576              ... bench:      53,173 ns/iter (+/- 644) = 19720 MB/s
test memcpy_rust_1048576_misalign     ... bench:      58,352 ns/iter (+/- 5,939) = 17969 MB/s
test memcpy_rust_1048576_offset       ... bench:      52,561 ns/iter (+/- 1,950) = 19949 MB/s
test memcpy_rust_4096                 ... bench:          84 ns/iter (+/- 20) = 48761 MB/s
test memcpy_rust_4096_misalign        ... bench:          96 ns/iter (+/- 2) = 42666 MB/s
test memcpy_rust_4096_offset          ... bench:          97 ns/iter (+/- 0) = 42226 MB/s
```

without branch:
```
test memcpy_rust_1048576              ... bench:      55,051 ns/iter (+/- 4,696) = 19047 MB/s
test memcpy_rust_1048576_misalign     ... bench:      57,791 ns/iter (+/- 545) = 18144 MB/s
test memcpy_rust_1048576_offset       ... bench:      53,902 ns/iter (+/- 1,893) = 19453 MB/s
test memcpy_rust_4096                 ... bench:          89 ns/iter (+/- 0) = 46022 MB/s
test memcpy_rust_4096_misalign        ... bench:          97 ns/iter (+/- 1) = 42226 MB/s
test memcpy_rust_4096_offset          ... bench:          97 ns/iter (+/- 0) = 42226 MB/s
```

It probably makes more sense to leave it out though.